gpio/src/gpio.rs

```diff
@@ -0,0 +1,947 @@
+            self.line_state.write().unwrap()[gpio as usize].val = match dir as u8 {
+                VIRTIO_GPIO_DIRECTION_NONE => None,
+                VIRTIO_GPIO_DIRECTION_IN => self.line_state.read().unwrap()[gpio as usize].val,
+                VIRTIO_GPIO_DIRECTION_OUT => Some(value as u16),
+
+                _ => return Err(Error::GpioDirectionInvalid(dir as u32)),
+            };
+
+            Ok(())
+        }
+
+        fn get_value(&self, gpio: u16) -> Result<u8> {
+            if self.get_value_result.is_err() {
+                return self.get_value_result;
+            }
+
+            if let Some(val) = self.line_state.read().unwrap()[gpio as usize].val {
+                Ok(val as u8)
+            } else {
+                Err(Error::GpioCurrentValueInvalid)
+            }
+        }
+
+        fn set_value(&self, gpio: u16, value: u32) -> Result<()> {
+            if self.set_value_result.is_err() {
+                return self.set_value_result;
+            }
+
+            self.line_state.write().unwrap()[gpio as usize].val = Some(value as u16);
+            Ok(())
+        }
+
+        fn set_irq_type(&self, _gpio: u16, _value: u16) -> Result<()> {
+            if self.set_irq_type_result.is_err() {
+                return self.set_irq_type_result;
+            }
+
+            Ok(())
+        }
+
+        fn wait_for_interrupt(&self, _gpio: u16) -> Result<()> {
+            if self.wait_for_irq_result.is_err() {
+                return self.wait_for_irq_result;
+            }
+
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_verify_gpio_controller() {
+        const NGPIO: u16 = 8;
+        let gpio_names = vec![
+            "gpio0".to_string(),
+            '\0'.to_string(),
+            "gpio2".to_string(),
+            '\0'.to_string(),
+            "gpio4".to_string(),
+            '\0'.to_string(),
+            "gpio6".to_string(),
+            '\0'.to_string(),
+        ];
+
+        // Controller adds '\0' for each line.
+        let names_size = size_of_val(&gpio_names) + gpio_names.len();
+
+        let mut device = DummyDevice::new(NGPIO);
+        device.gpio_names.clear();
+        device.gpio_names.append(&mut gpio_names.clone());
+        let controller = GpioController::new(device).unwrap();
+
+        assert_eq!(
+            *controller.get_config(),
+            VirtioGpioConfig {
+                ngpio: From::from(NGPIO),
+                padding: From::from(0),
+                gpio_names_size: From::from(names_size as u32),
+            }
+        );
+
+        let mut name = String::with_capacity(gpio_names.len());
+        for i in gpio_names {
+            name.push_str(&i);
+            name.push('\0');
+        }
+
+        assert_eq!(controller.get_num_gpios(), NGPIO);
+
+        assert_eq!(
+            controller
+                .operation(VIRTIO_GPIO_MSG_GET_LINE_NAMES, 0, 0)
+                .unwrap(),
+            name.as_bytes()
```

Comment:
Humm... In Controller I get `VIRTIO_GPIO_MSG_GET_LINE_NAMES => self.line_names.as_bytes().to_vec(),` but here I get `names.as_bytes()`.  No doubt Rust is doing some automatic conversion in the background but it might be worth adding a `.to_vec()` here to keep reviewers sane.

Reply:
With today's conversations I didn't have time to review a lot of new code.  More to come tomorrow.

Reply:
> Humm... In Controller I get `VIRTIO_GPIO_MSG_GET_LINE_NAMES => self.line_names.as_bytes().to_vec(),` but here I get `names.as_bytes()`. No doubt Rust is doing some automatic conversion in the background but it might be worth adding a `.to_vec()` here to keep reviewers sane.

You meant `name.as_bytes().to_vec()` ? Sure, done.